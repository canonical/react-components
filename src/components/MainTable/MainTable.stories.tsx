import { useState } from "react";
import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import MainTable from "./MainTable";
import Row from "../Row";
import Col from "../Col";
import ContextualMenu from "../ContextualMenu";

const meta: Meta<typeof MainTable> = {
  component: MainTable,
  tags: ["autodocs"],

  argTypes: {
    headers: {
      control: {
        disable: true,
      },
    },

    rows: {
      control: {
        disable: true,
      },
    },
  },
};

export default meta;

type Story = StoryObj<typeof MainTable>;

export const Default: Story = {
  name: "Default",

  args: {
    headers: [
      {
        content: null,
      },
      {
        content: "Foundation Cloud",
      },
      {
        content: "Foundation Cloud Plus",
      },
    ],

    rows: [
      {
        columns: [
          {
            content: "Expert delivery of an Ubuntu OpenStack cloud",
            role: "rowheader",
          },
          {
            content: "Reference architecture",
          },
          {
            content: "Custom architecture",
          },
        ],
      },
      {
        columns: [
          {
            content: "Workshop and training",
            role: "rowheader",
          },
          {
            content: "2-days",
          },
          {
            content: "4-days",
          },
        ],
      },
      {
        columns: [
          {
            content: "One-time price",
            role: "rowheader",
          },
          {
            content: "$75,000",
          },
          {
            content: "$150,000",
          },
        ],
      },
    ],
  },
};

export const Sortable: Story = {
  render: () => (
    <MainTable
      headers={[
        {
          content: "Status",
          sortKey: "status",
        },
        {
          content: "Cores",
          sortKey: "cores",
          className: "u-align--right",
        },
        {
          content: "RAM",
          sortKey: "ram",
          className: "u-align--right",
        },
        {
          content: "Disks",
          sortKey: "disks",
          className: "u-align--right",
        },
        {
          content: "Networks",
          className: "u-align--right",
        },
      ]}
      rows={[
        {
          columns: [
            {
              content: "Ready",
              role: "rowheader",
            },
            {
              content: 1,
              className: "u-align--right",
            },
            {
              content: "1 GiB",
              className: "u-align--right",
            },
            {
              content: 2,
              className: "u-align--right",
            },
            {
              content: 42,
              className: "u-align--right",
            },
          ],

          sortData: {
            status: "ready",
            cores: 2,
            ram: 1,
            disks: 2,
          },
        },
        {
          columns: [
            {
              content: "Idle",
              role: "rowheader",
            },
            {
              content: 1,
              className: "u-align--right",
            },
            {
              content: "1 GiB",
              className: "u-align--right",
            },
            {
              content: 2,
              className: "u-align--right",
            },
            {
              content: 23,
              className: "u-align--right",
            },
          ],

          sortData: {
            status: "idle",
            cores: 1,
            ram: 1,
            disks: 2,
          },
        },
        {
          columns: [
            {
              content: "Waiting",
              role: "rowheader",
            },
            {
              content: 8,
              className: "u-align--right",
            },
            {
              content: "3.9 GiB",
              className: "u-align--right",
            },
            {
              content: 3,
              className: "u-align--right",
            },
            {
              content: 0,
              className: "u-align--right",
            },
          ],

          sortData: {
            status: "waiting",
            cores: 8,
            ram: 3.9,
            disks: 3,
          },
        },
      ]}
      sortable
    />
  ),

  name: "Sortable",
};

export const Expanding: Story = {
  render: () => {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    let [expandedRow, setExpandedRow] = useState(1);

    return (
      <MainTable
        expanding
        headers={[
          {
            content: "Name",
          },
          {
            content: "Mac address",
          },
          {
            content: "IP",
          },
          {
            content: "Rack",
          },
          {
            content: "Last seen",
          },
          {
            content: "Actions",
            className: "u-align--right",
          },
        ]}
        rows={[
          {
            columns: [
              {
                content: "Unknown",
                role: "rowheader",
              },
              {
                content: "2c:44:fd:80:3f:25",
              },
              {
                content: "10.249.0.1",
              },
              {
                content: "karura",
              },
              {
                content: "Thu, 25 Oct. 2018 13:55:21",
              },
              {
                content: (
                  <button
                    className="u-toggle"
                    onClick={() => setExpandedRow(1)}
                  >
                    Show
                  </button>
                ),
                className: "u-align--right",
              },
            ],

            expanded: expandedRow === 1,

            expandedContent: (
              <Row>
                <Col size={8}>
                  <h4>Expanding table cell</h4>
                  <p>
                    Lorem ipsum dolor sit amet, consectetur adipisicing elit.
                    Consequuntur cum dicta beatae nostrum eligendi similique
                    earum, dolorem fuga quis, sequi voluptates architecto ipsa
                    dolorum eaque rem expedita inventore voluptas odit
                    aspernatur alias molestias facere.
                  </p>
                </Col>
              </Row>
            ),
          },
          {
            columns: [
              {
                content: "Unknown",
                role: "rowheader",
              },
              {
                content: "52:54:00:3a:fe:e9",
              },
              {
                content: "172.16.99.191",
              },
              {
                content: "karura",
              },
              {
                content: "Wed, 3 Oct. 2018 23:08:06",
              },
              {
                content: (
                  <button
                    className="u-toggle"
                    onClick={() => setExpandedRow(2)}
                  >
                    Show
                  </button>
                ),
                className: "u-align--right",
              },
            ],

            expanded: expandedRow === 2,

            expandedContent: (
              <Row>
                <Col size={8}>
                  <h4>Expanding table cell</h4>
                  <p>
                    Lorem ipsum dolor sit amet, consectetur adipisicing elit.
                    Consequuntur cum dicta beatae nostrum eligendi similique
                    earum, dolorem fuga quis, sequi voluptates architecto ipsa
                    dolorum eaque rem expedita inventore voluptas odit
                    aspernatur alias molestias facere.
                  </p>
                </Col>
              </Row>
            ),
          },
          {
            columns: [
              {
                content: "Unknown",
                role: "rowheader",
              },
              {
                content: "52:54:00:74:c2:10",
              },
              {
                content: "172.16.99.192",
              },
              {
                content: "karura",
              },
              {
                content: "Wed, 17 Oct. 2018 12:18:18",
              },
              {
                content: (
                  <button
                    className="u-toggle"
                    onClick={() => setExpandedRow(3)}
                  >
                    Show
                  </button>
                ),
                className: "u-align--right",
              },
            ],

            expanded: expandedRow === 3,

            expandedContent: (
              <Row>
                <Col size={8}>
                  <h4>Expanding table cell</h4>
                  <p>
                    Lorem ipsum dolor sit amet, consectetur adipisicing elit.
                    Consequuntur cum dicta beatae nostrum eligendi similique
                    earum, dolorem fuga quis, sequi voluptates architecto ipsa
                    dolorum eaque rem expedita inventore voluptas odit
                    aspernatur alias molestias facere.
                  </p>
                </Col>
              </Row>
            ),
          },
        ]}
      />
    );
  },

  name: "Expanding",
};

/**
 * If the table cell may have overflowing content (such as contextual menu) use `hasOverflow` prop to avoid cropping it.
 */
export const Overflow: Story = {
  render: () => {
    return (
      <MainTable
        headers={[
          {
            content: "Name",
          },
          {
            content: "Mac address",
          },
          {
            content: "IP",
          },
          {
            content: "Rack",
          },
          {
            content: "Last seen",
          },
          {
            content: "Actions",
            className: "u-align--right",
          },
        ]}
        rows={[
          {
            columns: [
              {
                content: "Unknown",
                role: "rowheader",
              },
              {
                content: "2c:44:fd:80:3f:25",
              },
              {
                content: "10.249.0.1",
              },
              {
                content: "karura",
              },
              {
                content: "Thu, 25 Oct. 2018 13:55:21",
              },
              {
                content: (
                  <ContextualMenu
                    links={[
                      {
                        children: "Link 1",
                        onClick: () => {},
                      },
                      {
                        children: "Link 2",
                        onClick: () => {},
                      },
                    ]}
                    hasToggleIcon
                    position="right"
                  />
                ),

                className: "u-align--right",
                hasOverflow: true,
              },
            ],
          },
          {
            columns: [
              {
                content: "Unknown",
                role: "rowheader",
              },
              {
                content: "52:54:00:3a:fe:e9",
              },
              {
                content: "172.16.99.191",
              },
              {
                content: "karura",
              },
              {
                content: "Wed, 3 Oct. 2018 23:08:06",
              },
              {
                content: (
                  <ContextualMenu
                    links={[
                      {
                        children: "Link 1",
                        onClick: () => {},
                      },
                      {
                        children: "Link 2",
                        onClick: () => {},
                      },
                    ]}
                    hasToggleIcon
                    position="right"
                  />
                ),

                className: "u-align--right",
                hasOverflow: true,
              },
            ],
          },
          {
            columns: [
              {
                content: "Unknown",
                role: "rowheader",
              },
              {
                content: "52:54:00:74:c2:10",
              },
              {
                content: "172.16.99.192",
              },
              {
                content: "karura",
              },
              {
                content: "Wed, 17 Oct. 2018 12:18:18",
              },
              {
                content: (
                  <ContextualMenu
                    links={[
                      {
                        children: "Link 1",
                        onClick: () => {},
                      },
                      {
                        children: "Link 2",
                        onClick: () => {},
                      },
                    ]}
                    hasToggleIcon
                    position="right"
                  />
                ),

                className: "u-align--right",
                hasOverflow: true,
              },
            ],
          },
        ]}
      />
    );
  },

  name: "Overflow",
};

export const Responsive: Story = {
  render: () => (
    <MainTable
      headers={[
        {
          content: (
            <span>
              Name<i className="p-icon--information"></i>
            </span>
          ),

          heading: "Name",
        },
        {
          content: "Users",
          className: "u-align--right",
        },
        {
          content: "Units",
          className: "u-align--right",
        },
      ]}
      rows={[
        {
          columns: [
            {
              content: "Ready",
              role: "rowheader",
            },
            {
              content: 1,
              className: "u-align--right",
            },
            {
              content: "1 GiB",
              className: "u-align--right",
            },
          ],
        },
        {
          columns: [
            {
              content: "Ready",
              role: "rowheader",
            },
            {
              content: 1,
              className: "u-align--right",
            },
            {
              content: "1 GiB",
              className: "u-align--right",
            },
          ],
        },
        {
          columns: [
            {
              content: "Ready",
              role: "rowheader",
            },
            {
              content: 1,
              className: "u-align--right",
            },
            {
              content: "1 GiB",
              className: "u-align--right",
            },
          ],
        },
      ]}
      responsive
    />
  ),

  name: "Responsive",
};

export const Paginated: Story = {
  render: () => (
    <MainTable
      headers={[
        {
          content: "Status",
        },
        {
          content: "Cores",
        },
      ]}
      paginate={5}
      rows={[
        {
          columns: [
            {
              content: "Ready",
            },
            {
              content: 1,
            },
          ],
        },
        {
          columns: [
            {
              content: "Idle",
            },
            {
              content: 4,
            },
          ],
        },
        {
          columns: [
            {
              content: "Waiting",
            },
            {
              content: 2,
            },
          ],
        },
        {
          columns: [
            {
              content: "Ready",
            },
            {
              content: 1,
            },
          ],
        },
        {
          columns: [
            {
              content: "Idle",
            },
            {
              content: 4,
            },
          ],
        },
        {
          columns: [
            {
              content: "Waiting",
            },
            {
              content: 2,
            },
          ],
        },
        {
          columns: [
            {
              content: "Ready",
            },
            {
              content: 1,
            },
          ],
        },
        {
          columns: [
            {
              content: "Idle",
            },
            {
              content: 4,
            },
          ],
        },
        {
          columns: [
            {
              content: "Waiting",
            },
            {
              content: 2,
            },
          ],
        },
        {
          columns: [
            {
              content: "Ready",
            },
            {
              content: 1,
            },
          ],
        },
        {
          columns: [
            {
              content: "Idle",
            },
            {
              content: 4,
            },
          ],
        },
        {
          columns: [
            {
              content: "Waiting",
            },
            {
              content: 2,
            },
          ],
        },
        {
          columns: [
            {
              content: "Ready",
            },
            {
              content: 1,
            },
          ],
        },
        {
          columns: [
            {
              content: "Idle",
            },
            {
              content: 4,
            },
          ],
        },
        {
          columns: [
            {
              content: "Waiting",
            },
            {
              content: 2,
            },
          ],
        },
        {
          columns: [
            {
              content: "Ready",
            },
            {
              content: 1,
            },
          ],
        },
        {
          columns: [
            {
              content: "Idle",
            },
            {
              content: 4,
            },
          ],
        },
        {
          columns: [
            {
              content: "Waiting",
            },
            {
              content: 2,
            },
          ],
        },
      ]}
    />
  ),

  name: "Paginated",
};

export const Empty: Story = {
  render: () => (
    <MainTable
      headers={[
        {
          content: "Plan",
        },
        {
          content: "Foundation Cloud",
        },
        {
          content: "Foundation Cloud Plus",
        },
      ]}
      rows={[]}
      emptyStateMsg="No data to display"
    />
  ),

  name: "Empty",
};
