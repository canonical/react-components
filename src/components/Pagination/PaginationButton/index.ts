export { default } from "./PaginationButton";
export type { Props as PaginationButtonProps } from "./PaginationButton";
