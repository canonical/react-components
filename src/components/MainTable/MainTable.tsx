import React, { useEffect, useMemo, useState } from "react";
import type { HTMLProps, ReactNode } from "react";

import type { ClassName, PropsWithSpread, SortDirection } from "types";
import Pagination from "../Pagination";
import Table from "../Table";
import type { TableProps } from "../Table";
import TableRow from "../TableRow";
import TableHeader from "../TableHeader";
import TableCell from "../TableCell";
import type { TableCellProps } from "../TableCell";
import { usePagination } from "hooks";

export type MainTableHeader = PropsWithSpread<
  {
    /**
     * The content of the table header.
     */
    content?: ReactNode;
    /**
     * Optional classes to apply to the table header.
     */
    className?: ClassName;
    /**
     * A key to sort the rows by. It should match a key given to the row `sortData`.
     */
    sortKey?: string | null;
    /**
     * Replacement value for data-heading if content is not a string.
     */
    heading?: string;
  },
  HTMLProps<HTMLTableHeaderCellElement>
>;

export type MainTableCell = PropsWithSpread<
  {
    /**
     * The content of the table cell.
     */
    content?: ReactNode;
  },
  // We explicitly omit "children" otherwise it's possible to overwrite "content".
  // Might want to consider just using "children" instead.
  Omit<TableCellProps, "children">
>;

export type MainTableRow = PropsWithSpread<
  {
    /**
     * Optional class(es) to apply to the row.
     */
    className?: ClassName;
    /**
     * The columns in this row.
     */
    columns?: MainTableCell[];
    /**
     * Whether this row should display as expanded.
     */
    expanded?: boolean;
    /**
     * The content to display when this column is expanded.
     */
    expandedContent?: ReactNode;
    /**
     * An optional key to identify this table row.
     */
    key?: number | string | null;
    /**
     * An object of data for use when sorting the rows. The keys of this object
     * should match the header sort keys.
     */
    sortData?: Record<string, unknown>;
  },
  HTMLProps<HTMLTableRowElement>
>;

export type Props = PropsWithSpread<
  {
    /**
     * The default key to sort the rows by.
     */
    defaultSort?: MainTableHeader["sortKey"];
    /**
     * The default direction the row data should be sorted by.
     */
    defaultSortDirection?: SortDirection;
    /**
     * A message to display when there are no table rows.
     */
    emptyStateMsg?: ReactNode;
    /**
     * The header columns for this table.
     */
    headers?: MainTableHeader[];
    /**
     * A function that is called when the sort key is changed.
     */
    onUpdateSort?: (sortKey: MainTableHeader["sortKey"]) => void;
    /**
     * A number of rows to paginate by.
     */
    paginate?: number | null;
    /**
     * The rows to display in the table.
     */
    rows?: MainTableRow[];
    /**
     * Whether this table should be sortable.
     */
    sortable?: boolean;
    /**
     * A function to be used when sorting.
     */
    sortFunction?: (
      a: MainTableRow,
      b: MainTableRow,
      currentSortDirection: SortDirection,
      currentSortKey: MainTableHeader["sortKey"]
    ) => -1 | 0 | 1;
    /**
     * A hidden caption to display on the table for screen readers
     */
    hiddenCaption?: string | null;
  },
  TableProps
>;

const updateSort = (
  setSortKey: (sortKey: MainTableHeader["sortKey"]) => void,
  setSortDirection: (direction: SortDirection) => void,
  sortKey: MainTableHeader["sortKey"],
  sortDirection: SortDirection
) => {
  let newDirection = null;
  if (sortDirection === "none") {
    newDirection = "ascending";
  } else if (sortDirection === "ascending") {
    newDirection = "descending";
  } else {
    sortKey = null;
  }
  setSortKey(sortKey);
  setSortDirection(newDirection);
};

const generateHeaders = (
  currentSortKey: MainTableHeader["sortKey"],
  currentSortDirection: SortDirection,
  expanding: Props["expanding"],
  headers: Props["headers"],
  sortable: Props["sortable"],
  setSortKey: (sortKey: MainTableHeader["sortKey"]) => void,
  setSortDirection: (direction: SortDirection) => void
) => {
  const headerItems = headers.map(({ content, sortKey, ...props }, index) => {
    let sortDirection: SortDirection;
    if (sortable && sortKey) {
      if (currentSortKey === sortKey) {
        sortDirection = currentSortDirection;
      } else {
        sortDirection = "none";
      }
    }
    return (
      <TableHeader
        key={index}
        sort={sortDirection}
        onClick={
          sortable && sortKey
            ? updateSort.bind(
                this,
                setSortKey,
                setSortDirection,
                sortKey,
                sortDirection
              )
            : undefined
        }
        {...props}
      >
        {content}
      </TableHeader>
    );
  });
  // When there is expanding content then provide an extra hidden header to
  // account for the extra cell in the body rows.
  return (
    <thead>
      <TableRow>
        {headerItems}
        {expanding && <TableHeader aria-hidden="true"></TableHeader>}
      </TableRow>
    </thead>
  );
};

const generateRows = ({
  rows,
  headers,
  responsive,
  expanding,
}: Required<Pick<Props, "rows">> &
  Pick<Props, "headers" | "responsive" | "expanding">) =>
  rows.map(
    (
      { columns, expanded, expandedContent, key, sortData, ...rowProps },
      index
    ) => {
      const cellItems = columns?.map(({ content, ...cellProps }, index) => {
        const headerContent = headers && headers[index]["content"];
        const headerReplacement = headers && headers[index]["heading"];

        if (responsive) {
          cellProps["data-heading"] =
            typeof headerContent === "string"
              ? headerContent
              : headerReplacement;
        }

        return (
          <TableCell key={index} {...cellProps}>
            {content}
          </TableCell>
        );
      });
      // if key was not provided as a prop, use row's index instead
      if (key === null || typeof key === "undefined") {
        key = index;
      }

      // The expanding cell is alway created to match the correct number of
      // table cells in rows that do have expanding content.
      return (
        <TableRow key={key} {...rowProps}>
          {cellItems}
          {expanding && (
            <TableCell expanding={true} hidden={!expanded}>
              {expandedContent}
            </TableCell>
          )}
        </TableRow>
      );
    }
  );

const sortRows = ({
  currentSortDirection,
  currentSortKey,
  rows,
  sortable,
  sortFunction,
}: Pick<Props, "rows" | "sortable" | "sortFunction"> & {
  currentSortDirection: Props["defaultSortDirection"];
  currentSortKey: Props["defaultSort"];
}): MainTableRow[] => {
  if (!rows) {
    return [];
  }
  // Clone the rows so we can restore the original order.
  const sortedRows = [...rows];
  if (sortable && currentSortKey) {
    if (!sortFunction) {
      sortFunction = (a, b) => {
        if (!a.sortData || !b.sortData) {
          return 0;
        }
        if (a.sortData[currentSortKey] > b.sortData[currentSortKey]) {
          return currentSortDirection === "ascending" ? 1 : -1;
        } else if (a.sortData[currentSortKey] < b.sortData[currentSortKey]) {
          return currentSortDirection === "ascending" ? -1 : 1;
        }
        return 0;
      };
    }
    sortedRows.sort((a, b) =>
      sortFunction(a, b, currentSortDirection, currentSortKey)
    );
  }
  return sortedRows;
};

/**
 * This is a [React](https://reactjs.org/) component to support many table use cases.
 */
const MainTable = ({
  defaultSort,
  defaultSortDirection,
  emptyStateMsg = "",
  expanding,
  headers,
  onUpdateSort,
  paginate,
  rows,
  responsive,
  sortable,
  sortFunction,
  hiddenCaption,
  ...props
}: Props): JSX.Element => {
  const [currentSortKey, setSortKey] = useState(defaultSort);
  const [currentSortDirection, setSortDirection] =
    useState(defaultSortDirection);

  // Update the current sort state if the prop changes.
  useEffect(() => {
    setSortKey(defaultSort);
  }, [defaultSort]);

  // Update the current sort direction state if the prop changes.
  useEffect(() => {
    setSortDirection(defaultSortDirection);
  }, [defaultSortDirection]);

  const updateSort = (newSort: MainTableHeader["sortKey"]) => {
    setSortKey(newSort);
    onUpdateSort && onUpdateSort(newSort);
  };

  const sortedRows = useMemo(
    () =>
      sortRows({
        currentSortDirection,
        currentSortKey,
        rows,
        sortable,
        sortFunction,
      }),
    [currentSortDirection, currentSortKey, rows, sortable, sortFunction]
  );

  const {
    pageData: finalRows,
    currentPage,
    paginate: setCurrentPage,
  } = usePagination(sortedRows, {
    itemsPerPage: paginate,
    autoResetPage: true,
  });

  return (
    <>
      <Table expanding={expanding} responsive={responsive} {...props}>
        {hiddenCaption && (
          <caption
            data-testid="hidden-caption"
            style={{
              height: "1px",
              left: "-1000px",
              overflow: "hidden",
              position: "absolute",
              top: "auto",
              width: "1px",
            }}
          >
            {hiddenCaption}
          </caption>
        )}
        {!!headers &&
          generateHeaders(
            currentSortKey,
            currentSortDirection,
            expanding,
            headers,
            sortable,
            updateSort,
            setSortDirection
          )}
        {
          // If the table has no rows, return empty state message
          Object.entries(finalRows).length === 0 && emptyStateMsg ? (
            <caption>{emptyStateMsg}</caption>
          ) : (
            <tbody>
              {generateRows({
                rows: finalRows,
                headers,
                responsive,
                expanding,
              })}
            </tbody>
          )
        }
      </Table>
      {paginate && rows && rows.length > 0 && (
        <Pagination
          currentPage={currentPage}
          itemsPerPage={paginate}
          paginate={setCurrentPage}
          style={{ marginTop: "1rem" }}
          totalItems={rows.length}
        />
      )}
    </>
  );
};

export default MainTable;
