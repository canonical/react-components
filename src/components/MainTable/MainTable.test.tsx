import { render, screen, within } from "@testing-library/react";
import React from "react";

import MainTable from "./MainTable";
import type { MainTableHeader, MainTableRow } from "./MainTable";
import { Label as PaginationButtonLabel } from "../Pagination/PaginationButton/PaginationButton";
import userEvent from "@testing-library/user-event";

describe("MainTable", () => {
  let headers: MainTableHeader[];
  let rows: MainTableRow[];

  beforeEach(() => {
    headers = [
      { content: "Status" },
      { content: "Cores", className: "u-align--right" },
      { content: "RAM", className: "u-align--right" },
      {
        content: (
          <span>
            Disks <i className="p-icon--information"></i>
          </span>
        ),
        className: "u-align--right",
      },
    ];
    rows = [
      {
        columns: [
          { content: "Ready", role: "rowheader" },
          { content: 1, className: "u-align--right" },
          { content: "1 GiB", className: "u-align--right" },
          { content: 2, className: "u-align--right" },
        ],
      },
      {
        columns: [
          { content: "Waiting", role: "rowheader" },
          { content: 1, className: "u-align--right" },
          { content: "1 GiB", className: "u-align--right" },
          { content: 2, className: "u-align--right" },
        ],
      },
      {
        columns: [
          { content: "Idle", role: "rowheader" },
          { content: 8, className: "u-align--right" },
          { content: "3.9 GiB", className: "u-align--right" },
          { content: 3, className: "u-align--right" },
        ],
      },
    ];
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it("renders", () => {
    render(<MainTable headers={headers} rows={rows} />);
    expect(screen.getByRole("grid")).toMatchSnapshot();
  });

  it("renders rows with provided keys", () => {
    // There isn't a way to access the key using RTL so this test checks for the
    // error: "Encountered two children with the same key".
    // This test helps to prevent this recurring issue, but if you're reading
    // this and know of a better way to write this test then please update it!
    const errorSpy = jest.spyOn(console, "error");
    rows[0].key = 1;
    rows[1].key = 2;
    rows[2].key = 0; // we explicitly want to test 0 as a falsy value provided as a key
    render(<MainTable headers={headers} rows={rows} />);
    expect(errorSpy).not.toHaveBeenCalled();
    errorSpy.mockReset();
  });

  it("can be expanding", () => {
    rows.push({
      columns: [
        { content: "Expanding", role: "rowheader" },
        { content: 1, className: "u-align--right" },
        { content: "1.9 GiB", className: "u-align--right" },
        { content: 2, className: "u-align--right" },
      ],
      expanded: true,
      expandedContent: <div>Expand this</div>,
    });
    render(<MainTable expanding={true} headers={headers} rows={rows} />);
    expect(screen.getByRole("grid")).toHaveClass("p-table--expanding");
    // Need to query the DOM as this cell is not exposed by default.
    // eslint-disable-next-line testing-library/no-node-access
    const heads = document.querySelectorAll("th");
    // There should be an additional hidden table header to account for the
    // expanding cell
    expect(heads.length).toEqual(headers.length + 1);
    expect(heads[heads.length - 1]).toHaveAttribute("aria-hidden", "true");
    // eslint-disable-next-line testing-library/no-node-access
    const columns = document.querySelectorAll("tr:last-child td");
    // There should be an additional table cell for the expanding content.
    expect(columns.length).toEqual(rows[rows.length - 1].columns.length + 1);
    const expandingColumn = columns[columns.length - 1];
    expect(expandingColumn.classList.contains("p-table__expanding-panel")).toBe(
      true
    );
    expect(expandingColumn.textContent).toBe("Expand this");
  });

  it("can be responsive", () => {
    render(<MainTable headers={headers} responsive={true} rows={rows} />);
    expect(screen.getByRole("grid")).toHaveClass("p-table--mobile-card");
  });

  it("contains data-heading attribute equal to heading on columns in responsive table", () => {
    render(<MainTable headers={headers} responsive={true} rows={rows} />);
    expect(screen.getAllByRole("rowheader")[0]).toHaveAttribute(
      "data-heading",
      "Status"
    );
  });

  it("doesn't contain data-heading attribute if there is no heading", () => {
    render(<MainTable responsive={true} rows={rows} />);
    expect(screen.getAllByRole("rowheader")[0]).not.toHaveAttribute(
      "data-heading"
    );
  });

  it("uses heading prop for data-heading if there are is no heading for that column", () => {
    headers[3].heading = "Replacement";
    render(<MainTable headers={headers} responsive={true} rows={rows} />);
    expect(screen.getAllByRole("gridcell")[2]).toHaveAttribute(
      "data-heading",
      "Replacement"
    );
  });

  it("can be paginated", () => {
    render(<MainTable paginate={2} rows={rows} />);
    expect(
      screen.getByRole("button", { name: PaginationButtonLabel.Next })
    ).toBeInTheDocument();
    expect(screen.getAllByRole("row")).toHaveLength(2);
  });

  it("can change the page", async () => {
    render(<MainTable paginate={2} rows={rows} />);
    await userEvent.click(
      screen.getByRole("button", { name: PaginationButtonLabel.Next })
    );
    expect(screen.getAllByRole("row")).toHaveLength(1);
    expect(screen.getAllByRole("rowheader")[0].textContent).toEqual("Idle");
  });

  describe("sorting", () => {
    beforeEach(() => {
      headers[0].sortKey = "status";
      headers[1].sortKey = "cores";
      headers[2].sortKey = "ram";
      rows[0].sortData = {
        status: "ready",
        cores: 2,
        ram: 1,
      };
      rows[1].sortData = {
        status: "waiting",
        cores: 1,
        ram: 1,
      };
      rows[2].sortData = {
        status: "idle",
        cores: 8,
        ram: 3.9,
      };
    });

    it("can be sortable", async () => {
      render(<MainTable headers={headers} rows={rows} sortable={true} />);
      // Sortable headers should have the sort prop set.
      expect(
        screen.getByRole("columnheader", { name: "Status" })
      ).toHaveAttribute("aria-sort", "none");
      // non-sortable headers should not have the sort prop set.
      expect(
        screen.getByRole("columnheader", { name: "Disks" })
      ).not.toHaveAttribute("aria-sort");
    });

    it("can sort when clicking on a header", async () => {
      render(<MainTable headers={headers} rows={rows} sortable={true} />);
      const rowItems = screen.getAllByRole("row");
      // Check the initial status order.
      expect(within(rowItems[1]).getByRole("rowheader").textContent).toBe(
        "Ready"
      );

      expect(within(rowItems[2]).getByRole("rowheader").textContent).toBe(
        "Waiting"
      );

      expect(within(rowItems[3]).getByRole("rowheader").textContent).toBe(
        "Idle"
      );

      await userEvent.click(
        screen.getByRole("columnheader", { name: "Status" })
      );
      // The status should now be ascending.
      expect(within(rowItems[1]).getByRole("rowheader").textContent).toBe(
        "Idle"
      );

      expect(within(rowItems[2]).getByRole("rowheader").textContent).toBe(
        "Ready"
      );

      expect(within(rowItems[3]).getByRole("rowheader").textContent).toBe(
        "Waiting"
      );

      await userEvent.click(
        screen.getByRole("columnheader", { name: "Status" })
      );
      // The status should now be descending.
      expect(within(rowItems[1]).getByRole("rowheader").textContent).toBe(
        "Waiting"
      );

      expect(within(rowItems[2]).getByRole("rowheader").textContent).toBe(
        "Ready"
      );

      expect(within(rowItems[3]).getByRole("rowheader").textContent).toBe(
        "Idle"
      );

      await userEvent.click(
        screen.getByRole("columnheader", { name: "Status" })
      );
      // The status be back to the original order.
      expect(within(rowItems[1]).getByRole("rowheader").textContent).toBe(
        "Ready"
      );

      expect(within(rowItems[2]).getByRole("rowheader").textContent).toBe(
        "Waiting"
      );

      expect(within(rowItems[3]).getByRole("rowheader").textContent).toBe(
        "Idle"
      );
    });

    it("can set a default sort", () => {
      render(
        <MainTable
          defaultSort="status"
          defaultSortDirection="descending"
          headers={headers}
          rows={rows}
          sortable={true}
        />
      );
      // Sortable headers should have the sort prop set.
      expect(
        screen.getByRole("columnheader", { name: "Status" })
      ).toHaveAttribute("aria-sort", "descending");
    });

    it("updates sort when props change", () => {
      const { rerender } = render(
        <MainTable
          defaultSort="status"
          defaultSortDirection="descending"
          headers={headers}
          rows={rows}
          sortable={true}
        />
      );
      expect(
        screen.getByRole("columnheader", { name: "Status" })
      ).toHaveAttribute("aria-sort", "descending");
      expect(
        screen.getByRole("columnheader", { name: "Cores" })
      ).toHaveAttribute("aria-sort", "none");
      rerender(
        <MainTable
          defaultSort="cores"
          defaultSortDirection="ascending"
          headers={headers}
          rows={rows}
          sortable={true}
        />
      );
      expect(
        screen.getByRole("columnheader", { name: "Status" })
      ).toHaveAttribute("aria-sort", "none");
      expect(
        screen.getByRole("columnheader", { name: "Cores" })
      ).toHaveAttribute("aria-sort", "ascending");
    });
  });
});
