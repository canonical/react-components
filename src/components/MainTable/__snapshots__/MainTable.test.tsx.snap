// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainTable renders 1`] = `
<table
  class=""
  role="grid"
>
  <thead>
    <tr
      role="row"
    >
      <th
        role="columnheader"
      >
        Status
      </th>
      <th
        class="u-align--right"
        role="columnheader"
      >
        Cores
      </th>
      <th
        class="u-align--right"
        role="columnheader"
      >
        RAM
      </th>
      <th
        class="u-align--right"
        role="columnheader"
      >
        <span>
          Disks 
          <i
            class="p-icon--information"
          />
        </span>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      role="row"
    >
      <td
        aria-hidden="false"
        class=""
        role="rowheader"
      >
        Ready
      </td>
      <td
        aria-hidden="false"
        class="u-align--right"
        role="gridcell"
      >
        1
      </td>
      <td
        aria-hidden="false"
        class="u-align--right"
        role="gridcell"
      >
        1 GiB
      </td>
      <td
        aria-hidden="false"
        class="u-align--right"
        role="gridcell"
      >
        2
      </td>
    </tr>
    <tr
      role="row"
    >
      <td
        aria-hidden="false"
        class=""
        role="rowheader"
      >
        Waiting
      </td>
      <td
        aria-hidden="false"
        class="u-align--right"
        role="gridcell"
      >
        1
      </td>
      <td
        aria-hidden="false"
        class="u-align--right"
        role="gridcell"
      >
        1 GiB
      </td>
      <td
        aria-hidden="false"
        class="u-align--right"
        role="gridcell"
      >
        2
      </td>
    </tr>
    <tr
      role="row"
    >
      <td
        aria-hidden="false"
        class=""
        role="rowheader"
      >
        Idle
      </td>
      <td
        aria-hidden="false"
        class="u-align--right"
        role="gridcell"
      >
        8
      </td>
      <td
        aria-hidden="false"
        class="u-align--right"
        role="gridcell"
      >
        3.9 GiB
      </td>
      <td
        aria-hidden="false"
        class="u-align--right"
        role="gridcell"
      >
        3
      </td>
    </tr>
  </tbody>
</table>
`;
